python/ray/serve/http_proxy.py

```diff
@@ -10,62 +10,176 @@
+                LongPollNamespace.ROUTE_TABLE: self.router.update_routes,
+            },
+            call_in_event_loop=asyncio.get_event_loop())
+        self.request_counter = metrics.Counter(
+            "serve_num_http_requests",
+            description="The number of HTTP requests processed.",
+            tag_keys=("route", ))
+
+    async def block_until_endpoint_exists(self, endpoint: EndpointTag,
+                                          timeout_s: float):
+        start = time.time()
+        while True:
+            if time.time() - start > timeout_s:
+                raise TimeoutError(
+                    f"Waited {timeout_s} for {endpoint} to propagate.")
+            if self.router.endpoint_exists(endpoint):
+                return
+            await asyncio.sleep(0.2)
+
+    async def _not_found(self, scope, receive, send):
+        current_path = scope["path"]
+        response = Response(
+            f"Path '{current_path}' not found. "
+            "Please ping http://.../-/routes for route table.",
+            status_code=404)
+        await response.send(scope, receive, send)
 
     async def __call__(self, scope, receive, send):
-        http_body_bytes = await self.receive_http_body(scope, receive, send)
+        """Implements the ASGI protocol.
 
-        headers = {k.decode(): v.decode() for k, v in scope["headers"]}
+        See details at:
+            https://asgi.readthedocs.io/en/latest/specs/index.html.
+        """
+
+        assert scope["type"] == "http"
+        self.request_counter.inc(tags={"route": scope["path"]})
+
+        if scope["path"] == "/-/routes":
+            return await starlette.responses.JSONResponse(
+                self.router.route_info)(scope, receive, send)
 
-        # scope["router"] and scope["endpoint"] contain references to a router
-        # and endpoint object, respectively, which each in turn contain a
-        # reference to the Serve client, which cannot be serialized.
-        # The solution is to delete these from scope, as they will not be used.
-        del scope["router"]
-        del scope["endpoint"]
+        route_prefix, handle = self.router.match_route(scope["path"],
+                                                       scope["method"])
+        if route_prefix is None:
+            return await self._not_found(scope, receive, send)
+
+        http_body_bytes = await receive_http_body(scope, receive, send)
 
         # Modify the path and root path so that reverse lookups and redirection
         # work as expected. We do this here instead of in replicas so it can be
         # changed without restarting the replicas.
-        scope["path"] = scope["path"].replace(self.path_prefix, "", 1)
-        scope["root_path"] = self.path_prefix
-        handle = self.handle.options(
+        if route_prefix.endswith("/"):
+            # We don't want to remove the trailing '/' from the incoming path.
+            route_prefix = route_prefix[:-1]
```

Comment:
- This case doesn't exist anymore because it will never end with / unless root.
- And this case will fail with route-prefix == "/"?

Reply:
Ah yeah this is leftover from a previous iteration. Let me make sure that we're handling the root properly.